File: Easy/LC543-Diameter-of-Binary-Tree/code.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
    
        def height(root):
            if not root:
                return 0
            return 1 + max(height(root.left), height(root.right))
            
        def diameter(root):
            if not root:
                return 0
            left_dia = diameter(root.left)
            right_dia = diameter(root.right)
            dia = height(root.left) + height(root.right)
            return max([left_dia, right_dia, dia])
        return diameter(root)

File: Easy/LC543-Diameter-of-Binary-Tree/code.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int height(TreeNode* root) {
    if(root == NULL)
        return 0;
    else
        return 1 + max(height(root->left),height(root->right));
}
public:
    int diameterOfBinaryTree(TreeNode* root) {
        if(root == NULL) return 0;
        int curr_dia = height(root->left) + height(root->right);
        int left_subtree_dia = diameterOfBinaryTree(root->left);
        int right_subtree_dia = diameterOfBinaryTree(root->right);
        int mx = max(left_subtree_dia,right_subtree_dia);
        return max(curr_dia,mx);
    }
};
